### PR TITLE
Feat/provenance project node

### PR DESCRIPTION
**Pull request by marshHawk4**

### CODE REVIEW

File: askem.dbml [hunk 1]
```diff
@@ -15,11 +15,29 @@ Enum public.resource_type {
...
+   PlanParameter
    ModelRevision
    Intermediate
    Publication
    SimulationRun
+   Project
+   Concept
+}
+
+Enum provenance_abbreviations{
```
Comment (fivegrant) [inline askem.dbml:29]: Should this be in the `askem.dbml`? Isn't this part of the implementation?

Comment (fivegrant) [reply askem.dbml:29]: Couldn't it just be included here: https://github.com/DARPA-ASKEM/data-service/blob/2880f968cb64bb0809faf8b52a12e646227fcd4e/tds/schema/provenance.py#L21-L35

File: tds/routers/models.py [hunk 2]
```diff
@@ -324,115 +381,126 @@ def model_opt(
...
+            )
 
-            elif model_operation in ("decompose", "glue"):
-                state = orm.ModelState(content=payload.get("content"))
-            else:
-                raise TypeError("Operation not supported")
+        elif model_operation in ("decompose", "glue"):
+            state = orm.ModelState(content=payload.get("content"))
+        else:
+            raise HTTPException(status_code=404, detail="Operation not supported")
```
Comment (fivegrant) [inline tds/routers/models.py:402]:
```suggestion
            raise HTTPException(status_code=400, detail="Operation not supported")
```

File: askem.dbml [hunk 3]
```diff
@@ -15,11 +15,29 @@ Enum public.resource_type {
...
+  Pp
+  Mr
+  In
+  Pu
+  Sr
+  Pr
+  Cn
 }
 
 Enum provenance_search_types{
```
Comment (fivegrant) [inline askem.dbml:43]: Do we want this in the dbml?

File: scripts/provenance_fake_data.py [hunk 4]
```diff
@@ -353,66 +354,65 @@ def upload_fake_provanence_data(person_id=1, project_id=1):
...
+                                        "context", {}
+                                    ).get("property", None)
+                                    if context is not None:
+                                        print(f"adding concept context {context}")
+                                        add_concept(
+                                            concept=context,
+                                            object_id=parameter.get("id"),
+                                            type="simulation_parameters",
+                                        )
+                                except Exception as e:
```
Comment (fivegrant) [inline scripts/provenance_fake_data.py:368]: Universal exception. How are you expecting things here to fail? Are you expecting any of the lines to fail or one in particular? Could we be using `try..except..else` or `try...except...finally`?

File: tds/db/graph/query_helpers.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    allow for different relationship types
+    """
+    if direction == "all":
+        return f"-[{relationship_type}]-"
+    if direction == "child":
+        return f"<-[{relationship_type}]-"
+    if direction == "parent":
+        return f"-[{relationship_type}]->"
+    raise HTTPException(
+        status_code=404, detail="Relationship direction is not allowed."
```
Comment (fivegrant) [inline tds/db/graph/query_helpers.py:22]:
```suggestion
        status_code=400, detail="Relationship direction is not allowed."
```

File: tds/db/graph/query_helpers.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+        modelr_node = node_builder(node_type="ModelRevision")
+        return (
+            match_node
+            + f"-[r *1..]->{modelr_node}"
+            + "-[r2:REINTERPRETS]->"
+            + f"{root_node}"
+        )
+    raise HTTPException(
+        status_code=404, detail=f"Models can not be derived from this type: {root_type}"
```
Comment (fivegrant) [inline tds/db/graph/query_helpers.py:50]:
```suggestion
        status_code=400, detail=f"Models can not be derived from this type: {root_type}"
```

File: tds/db/graph/search_provenance.py [hunk 7]
```diff
@@ -62,60 +85,214 @@ def parent_nodes(self, payload):
...
-        Return models derived from a publication or intermediate
+        Return models derived from artifact (Publication or Intermediate)
         """
-        if root_type == "publication":
-            return (
-                "Match (m:model)-[r *1..]->(i:intermediate)-[r2:EXTRACTED_FROM]->"
-                + f"(n:{root_type})"
+        if payload.get("root_type") not in ("Publication", "Intermediate"):
+            raise HTTPException(
+                status_code=404,
```
Comment (fivegrant) [inline tds/db/graph/search_provenance.py:94]:
```suggestion
                status_code=400,
```

File: tds/db/graph/search_provenance.py [hunk 8]
```diff
@@ -62,60 +85,214 @@ def parent_nodes(self, payload):
...
+        latest model which was used to create the artifact
         """
+        if payload.get("root_type") not in (
+            "Model",
+            "SimulationRun",
+            "Plan",
+            "Dataset",
+        ):
+            raise HTTPException(
+                status_code=404,
```
Comment (fivegrant) [inline tds/db/graph/search_provenance.py:124]:
```suggestion
                status_code=400,
```

File: tds/db/graph/search_provenance.py [hunk 9]
```diff
@@ -62,60 +85,214 @@ def parent_nodes(self, payload):
...
-                + "RETURN labels(m) as label, m.id as id"
+                f"{match_pattern}"
+                + "Match (Mr2:ModelRevision)"
+                + f"-[r2:{relationships_str} *1.. ]->(Mr) "
+                + "With collect(Mr)+collect(Mr2) as Mrs "
+                + "Unwind Mrs as Both_rms "
+                + "With DISTINCT Both_rms "
+                + "RETURN labels(Both_rms) as label, Both_rms.id as id "
             )
+            print("heee")
```
Comment (fivegrant) [inline tds/db/graph/search_provenance.py]:
```suggestion
```

File: tds/db/graph/search_provenance.py [hunk 10]
```diff
@@ -62,60 +85,214 @@ def parent_nodes(self, payload):
...
             print(query)
```
Comment (fivegrant) [inline tds/db/graph/search_provenance.py]:
```suggestion
            logger.info(query)
```

File: tds/db/graph/search_provenance.py [hunk 11]
```diff
@@ -62,60 +85,214 @@ def parent_nodes(self, payload):
...
+            if len(response_data) == 0:
+                print("here")
+                query = f"{match_pattern}" + "RETURN labels(Mr) as label, Mr.id as id "
+                response = session.run(query)
+                response_data = [
+                    {res.data().get("label")[0]: res.data().get("id")}
+                    for res in response
+                ]
+
+            print(response_data)
```
Comment (fivegrant) [inline tds/db/graph/search_provenance.py:163]:
```suggestion
```

File: tds/routers/models.py [hunk 12]
```diff
@@ -324,115 +381,126 @@ def model_opt(
...
-            if model_operation == "copy":
-                state = orm.ModelState(
-                    content=session.query(orm.ModelState)
-                    .get(payload.get("left"))
-                    .__dict__.get("content")
-                )
+    with Session(rdb) as session:
+        payload = payload.dict()
+        # query old model and old content
+        # print(payload)
```
Comment (fivegrant) [inline tds/routers/models.py:387]:
```suggestion
```

File: tds/routers/models.py [hunk 13]
```diff
@@ -466,6 +534,7 @@ def update_model(
             model.state_id = state.id
             session.commit()
             if settings.NEO4J_ENABLED:
+                print("here")
```
Comment (fivegrant) [inline tds/routers/models.py:537]:
```suggestion
```

File: tds/routers/models.py [hunk 14]
```diff
@@ -218,14 +218,44 @@ def get_model_parameters(
...
 ) -> Response:
     """
     Update the parameters for a model
     """
     with Session(rdb) as session:
         adjust_model_params(id, payload, session)
         session.commit()
+
+    if settings.NEO4J_ENABLED:
+        print("hererer")
```
Comment (fivegrant) [inline tds/routers/models.py:234]:
```suggestion
```

File: tds/routers/models.py [hunk 15]
```diff
@@ -218,14 +218,44 @@ def get_model_parameters(
...
+            provenance_handler = ProvenanceHandler(rdb=rdb, graph_db=graph_db)
+
+            with Session(rdb) as session:
+                model = session.query(orm.ModelDescription).get(id)
+                parameters: Query[orm.ModelParameter] = session.query(
+                    orm.ModelParameter
+                ).filter(orm.ModelParameter.model_id == id)
+
+            updated_parameters = orm_to_params(list(parameters))
+            print(updated_parameters)
```
Comment (fivegrant) [inline tds/routers/models.py:246]:
```suggestion
```

File: tds/routers/models.py [hunk 16]
```diff
@@ -218,14 +218,44 @@ def get_model_parameters(
...
+            # add ModelParameter nodes
+            for parameter in updated_parameters:
+                print(parameter)
```
Comment (fivegrant) [inline tds/routers/models.py:249]:
```suggestion
```

File: tds/routers/models.py [hunk 17]
```diff
@@ -291,19 +321,46 @@ def create_model(
                 )
             )
         session.commit()
-        if settings.NEO4J_ENABLED:
-            logger.info("Neo4j is set")
-            provenance_handler = ProvenanceHandler(rdb=rdb, graph_db=graph_db)
+
+    if settings.NEO4J_ENABLED:
+        print("Neo4j is set")
```
Comment (fivegrant) [inline tds/routers/models.py:326]:
```suggestion
        logger.info("Neo4j is set")
```

File: tds/db/graph/search_provenance.py [hunk 18]
```diff
@@ -12,38 +26,47 @@ class SearchProvenance(ProvenanceHandler):
...
-                f"Match (n1: {payload.get('root_type')}) "
-                + f"{self.dynamic_relationship_direction(direction=direction)}(n2)"
-                + "Where n1.id = $root_id "
-                + "RETURN labels(n2) as label, n2.id as id"
+                f"{match_node}"
+                + f"{relation_direction}(n) "
+                + "With DISTINCT n "
+                + "RETURN labels(n) as label, n.id as id"
             )
+            print(query)
```
Comment (fivegrant) [inline tds/db/graph/search_provenance.py:59]:
```suggestion
```

File: tds/db/graph/search_provenance.py [hunk 19]
```diff
@@ -12,38 +26,47 @@ class SearchProvenance(ProvenanceHandler):
...
+            logging.info(query)
 
-            response = session.run(query, root_id=payload.get("root_id"))
+            response = session.run(query)
 
-            return [
-                {"label": res.data().get("label")[0], "id": res.data().get("id")}
-                for res in response
+            response_data = [
+                {res.data().get("label")[0]: res.data().get("id")} for res in response
```
Comment (fivegrant) [inline tds/db/graph/search_provenance.py:65]: What is `res`? Is it okay if this line is broken up. Just not sure what it's doing.

**Review by fivegrant — CHANGES_REQUESTED** [review]

**Comment by fivegrant** [comment]
rebase is needed because of this: https://github.com/DARPA-ASKEM/data-service/pull/143#pullrequestreview-1239324109

File: scripts/provenance_fake_data.py (hunk 4 above)
Comment (marshHawk4) [reply scripts/provenance_fake_data.py]: removed

File: askem.dbml (hunk 3 above)
Comment (marshHawk4) [reply askem.dbml:30]: I moved it out. Doesn't need to be there

File: tds/routers/models.py (hunk 2 above)
Comment (marshHawk4) [reply tds/routers/models.py]: Changed

File: tds/routers/models.py (hunk 15 above)
Comment (marshHawk4) [reply tds/routers/models.py]: removed

File: tds/db/graph/search_provenance.py (hunk 19 above)
Comment (marshHawk4) [reply tds/db/graph/search_provenance.py:64]: res is just part of the result that is looping. It is going to be changed anyways in my next pr to account for returning the whole graph.

File: tds/routers/models.py (hunk 17 above)
Comment (marshHawk4) [reply tds/routers/models.py]: updated

File: tds/db/graph/query_helpers.py (hunk 5 above)
Comment (marshHawk4) [reply tds/db/graph/query_helpers.py]: updated

File: tds/settings.py [hunk 20]
```diff
@@ -19,7 +19,7 @@ class Settings(BaseSettings):
     DKG_URL = "http://34.230.33.149"
     DKG_API_PORT = 8771
     DKG_DESC_PORT = 8772
-    NEO4J_ENABLED = True
+    NEO4J_ENABLED = False
```
Comment (fivegrant) [inline tds/settings.py:22]: Do we want to this to be False by default? `NEO4J_ENABLED` could be set to `True` in `api.env.sample` so we wouldn't have to change it here.

File: tds/db/graph/search_provenance.py [hunk 21]
```diff
@@ -62,60 +82,206 @@ def parent_nodes(self, payload):
...
-                + f"(n:{root_type})"
+        with self.graph_db.session() as session:
+
+            generated_query = derived_models_query_generater(
+                root_type=payload.get("root_type"), root_id=payload.get("root_id")
             )
-        raise Exception(f"Models can not be derived from this type: {root_type}")
 
-    def derived_models(self, payload):
+            print(generated_query)
```
Comment (fivegrant) [inline tds/db/graph/search_provenance.py:101]:
```suggestion
```

File: tds/db/graph/query_helpers.py [hunk 22]
```diff
@@ -0,0 +1,207 @@
...
+            "-[r3:EDITED_FROM|COPIED_FROM|GLUED_FROM|STRATIFED_FROM  *1..]-"
+            "(Mr2:ModelRevision) "
+            "with *,collect(r)+collect(r2)+collect(r3) as r4, "
+            "collect(Mr)+collect(Mr2) as ms "
+            "unwind ms as mss "
+            "unwind r4 as r5 "
+            "Optional Match(mss)<-[r6:BEGINS_AT]-(Md:Model) "
+            "with *, collect(r5)+collect(r6) as r7 "
+            "unwind r7 as r8 "
+            "return Pu, In, ms,Md, r8"
```
Comment (fivegrant) [inline tds/db/graph/query_helpers.py:46]:
```suggestion
            f"""
            Match(Pu:Publication {{id:{root_id}}})
            <-[r:EXTRACTED_FROM]-(In:Intermediate)
            Optional Match (In)<-[r2:REINTERPRETS *1..]-(Mr:ModelRevision)
            Optional Match(Mr)
            -[r3:EDITED_FROM|COPIED_FROM|GLUED_FROM|STRATIFED_FROM  *1..]-
            (Mr2:ModelRevision)
            with *,collect(r)+collect(r2)+collect(r3) as r4,
            collect(Mr)+collect(Mr2) as ms
            unwind ms as mss
            unwind r4 as r5
            Optional Match(mss)<-[r6:BEGINS_AT]-(Md:Model)
            with *, collect(r5)+collect(r6) as r7
            unwind r7 as r8 
            return Pu, In, ms,Md, r8
            """
```

File: tds/db/graph/query_helpers.py [hunk 23]
```diff
@@ -0,0 +1,207 @@
...
+            "-[r3:EDITED_FROM|COPIED_FROM|GLUED_FROM|STRATIFED_FROM  *1..]-"
+            "(Mr2:ModelRevision) "
+            "with *,collect(r2)+collect(r3) as r4, collect(Mr)+collect(Mr2) as ms "
+            "unwind ms as mss "
+            "unwind r4 as r5 "
+            "Optional Match(mss)<-[r6:BEGINS_AT]-(Md:Model) "
+            "with *, collect(r5)+collect(r6) as r7 "
+            "unwind r7 as r8 "
+            "return  In, ms,Md, r8"
+        )
```
Comment (fivegrant) [inline tds/db/graph/query_helpers.py:62]:
```suggestion
        return f"""
            Match (In:Intermediate {{id:{root_id}}})<-[r2:REINTERPRETS *1..]
            -(Mr:ModelRevision)
            Optional Match(Mr)
            -[r3:EDITED_FROM|COPIED_FROM|GLUED_FROM|STRATIFED_FROM  *1..]-
            (Mr2:ModelRevision)
            with *,collect(r2)+collect(r3) as r4, collect(Mr)+collect(Mr2) as ms
            unwind ms as mss
            unwind r4 as r5
            Optional Match(mss)<-[r6:BEGINS_AT]-(Md:Model)
            with *, collect(r5)+collect(r6) as r7
            unwind r7 as r8
            return  In, ms,Md, r8
            """
```

**Review by fivegrant — CHANGES_REQUESTED** [review]

File: tds/db/graph/search_provenance.py (hunk 21 above)
Comment (marshHawk4) [reply tds/db/graph/search_provenance.py:101]: removed

File: tds/db/graph/query_helpers.py (hunk 23 above)
Comment (marshHawk4) [reply tds/db/graph/query_helpers.py:62]: changed

File: tds/db/graph/search_provenance.py (hunk 10 above)
Comment (fivegrant) [reply tds/db/graph/search_provenance.py:147]: git is wrong

**Review by fivegrant — APPROVED** [review]
Works for me, but we should wait until @mattprintz to merge

**Review by mattprintz — APPROVED** [review]
Looks good pending testing/sign-off on the startup script changes I made.

**Pull request merged**